build/xtask/src/dist.rs

```diff
@@ -1712,198 +1711,156 @@ fn allocate_one(
+                flash
             );
-        }
-
-        // Generate a RegionDesc for each uniquely allocated memory region
-        // referenced by this task, and install them as entries 0..N in the
-        // task's region table.
-        let allocs = &task_allocations[name];
-        for (ri, (output_name, range)) in allocs.iter().enumerate() {
-            let region: Vec<&crate::config::Output> = toml.outputs[output_name]
-                .iter()
-                .filter(|o| o.name == *image_name)
-                .collect();
-
-            if region.len() > 1 {
-                bail!("Multiple regions defined for image {}", image_name);
-            }
-
-            let out = region[0];
-
-            let mut attributes = abi::RegionAttributes::empty();
-            if out.read {
-                attributes |= abi::RegionAttributes::READ;
-            }
-            if out.write {
-                attributes |= abi::RegionAttributes::WRITE;
-            }
-            if out.execute {
-                attributes |= abi::RegionAttributes::EXECUTE;
-            }
-            if out.dma {
-                attributes |= abi::RegionAttributes::DMA;
-            }
-            // no option for setting DEVICE for this region
-
-            task_regions[ri] = regions.len() as u8;
-
-            regions.push(abi::RegionDesc {
-                base: range.start,
-                size: range.end - range.start,
-                attributes,
-            });
-        }
+        };
 
-        // For peripherals referenced by the task, we don't need to allocate
-        // _new_ regions, since we did them all in advance. Just record the
-        // entries for the TaskDesc.
-        for (j, peripheral_name) in task.uses.iter().enumerate() {
-            if let Some(&peripheral) = peripheral_index.get(&peripheral_name) {
-                task_regions[allocs.len() + j] = peripheral as u8;
-            } else {
-                bail!(
-                    "Could not find peripheral `{}` referenced by task `{}`.",
-                    peripheral_name,
-                    name
-                );
-            }
+        // Mark off the regions this task uses.
+        for region in &task.uses {
+            used_shared_regions.insert(region);
         }
 
-        let mut flags = abi::TaskFlags::empty();
-        if task.start {
-            flags |= abi::TaskFlags::START_AT_BOOT;
-        }
+        // Prep this task's shared region name set.
+        let shared_regions: std::collections::BTreeSet<String> =
+            task.uses.iter().cloned().collect();
+
+        let owned_regions = task_allocations[name].iter()
+            .map(|(out_name, range)| {
+                // Look up region for this image
+                let mut regions = toml.outputs[out_name].iter()
+                    .filter(|o| &o.name == image_name);
+                let out = regions.next().expect("no region for name");
+                if regions.next().is_some() {
```

Comment:
Nit: it's a little inefficient to do this check repeatedly in the inner-most loop, versus checking `toml.outputs` once elsewhere.

Reply:
At this point, I have _no idea_ what invariants are being maintained on the toml datastructure (i.e. which checks I can rely upon having done elsewhere), so while I agree that it would be more efficient to "parse don't validate" this, I feel like we're kind of in debt here and I'm not sure where to do it.